source/common/jwt_authn/jwks.cc

```diff
@@ -0,0 +1,249 @@
+      return nullptr;
+    }
+    bssl::UniquePtr<BIGNUM> bn_x = createBigNumFromBase64UrlString(x);
+    bssl::UniquePtr<BIGNUM> bn_y = createBigNumFromBase64UrlString(y);
+    if (!bn_x || !bn_y) {
+      // EC public key field is missing or has parse error.
+      updateStatus(Status::JwksEcParseError);
+      return nullptr;
+    }
+
+    if (EC_KEY_set_public_key_affine_coordinates(ec_key.get(), bn_x.get(), bn_y.get()) == 0) {
+      updateStatus(Status::JwksEcParseError);
+      return nullptr;
+    }
+    return ec_key;
+  }
+
+private:
+  bssl::UniquePtr<EVP_PKEY> createEvpPkeyFromRsa(RSA* rsa) {
+    if (!rsa) {
+      return nullptr;
+    }
+    bssl::UniquePtr<EVP_PKEY> key(EVP_PKEY_new());
+    EVP_PKEY_set1_RSA(key.get(), rsa);
+    return key;
+  }
+
+  bssl::UniquePtr<BIGNUM> createBigNumFromBase64UrlString(const std::string& s) {
+    std::string s_decoded = decodeBase64Url(s);
+    if (s_decoded == "") {
+      return nullptr;
+    }
+    return bssl::UniquePtr<BIGNUM>(BN_bin2bn(castToUChar(s_decoded), s_decoded.length(), NULL));
+  };
+
+  bssl::UniquePtr<RSA> createRsaFromJwk(const std::string& n, const std::string& e) {
+    bssl::UniquePtr<RSA> rsa(RSA_new());
+    // It crash if RSA object couldn't be created.
+    ASSERT(rsa);
+
+    rsa->n = createBigNumFromBase64UrlString(n).release();
+    rsa->e = createBigNumFromBase64UrlString(e).release();
+    if (!rsa->n || !rsa->e) {
+      // RSA public key field is missing or has parse error.
+      updateStatus(Status::JwksRsaParseError);
+      return nullptr;
+    }
+    return rsa;
+  }
+};
+
+} // namespace
+
+JwksPtr Jwks::createFrom(const std::string& pkey, Type type) {
+  JwksPtr keys(new Jwks());
+  switch (type) {
+  case Type::JWKS:
+    keys->createFromJwksCore(pkey);
+    break;
+  case Type::PEM:
+    keys->createFromPemCore(pkey);
+    break;
+  default:
+    PANIC("can not reach here");
+  }
+  return keys;
+}
+
+void Jwks::createFromPemCore(const std::string& pkey_pem) {
+  keys_.clear();
+  PubkeyPtr key_ptr(new Pubkey());
+  EvpPkeyGetter e;
+  key_ptr->evp_pkey_ = e.createEvpPkeyFromStr(pkey_pem);
+  key_ptr->pem_format_ = true;
+  updateStatus(e.getStatus());
+  if (e.getStatus() == Status::Ok) {
+    keys_.push_back(std::move(key_ptr));
+  }
+}
+
+void Jwks::createFromJwksCore(const std::string& pkey_jwks) {
+  keys_.clear();
+
+  Json::ObjectSharedPtr jwks_json;
+  try {
+    jwks_json = Json::Factory::loadFromString(pkey_jwks);
+  } catch (Json::Exception& e) {
+    updateStatus(Status::JwksParseError);
+    return;
+  }
+  std::vector<Json::ObjectSharedPtr> keys;
+  if (!jwks_json->hasObject("keys")) {
+    updateStatus(Status::JwksNoKeys);
+    return;
+  }
+  try {
+    keys = jwks_json->getObjectArray("keys", true);
+  } catch (Json::Exception& e) {
+    updateStatus(Status::JwksBadKeys);
+    return;
+  }
+
+  for (auto jwk_json : keys) {
+    try {
+      extractJwk(jwk_json);
+    } catch (Json::Exception& e) {
+      continue;
```

Comment:
no need continue, log the exception here at least?

Reply:
Done